pkg/services/authn/clients/ext_jwt.go

```diff
@@ -0,0 +1,226 @@
+package clients
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-jose/go-jose/v3/jwt"
+	"golang.org/x/exp/slices"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/authn"
+	"github.com/grafana/grafana/pkg/services/signingkeys"
+	"github.com/grafana/grafana/pkg/services/user"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+var _ authn.Client = new(ExtendedJWT)
+
+var (
+	acceptedSigningMethods = []string{"RS256", "ES256"}
+	timeNow                = time.Now
+)
+
+const (
+	rfc9068ShortMediaType = "at+jwt"
+	rfc9068MediaType      = "application/at+jwt"
+)
+
+func ProvideExtendedJWT(userService user.Service, cfg *setting.Cfg, signingKeys signingkeys.Service) *ExtendedJWT {
+	return &ExtendedJWT{
+		cfg:         cfg,
+		log:         log.New(authn.ClientExtendedJWT),
+		userService: userService,
+		signingKeys: signingKeys,
+	}
+}
+
+type ExtendedJWT struct {
+	cfg         *setting.Cfg
+	log         log.Logger
+	userService user.Service
+	signingKeys signingkeys.Service
+}
+
+type ExtendedJWTClaims struct {
+	jwt.Claims
+	ClientID     string              `json:"client_id"`
+	Groups       []string            `json:"groups"`
+	Email        string              `json:"email"`
+	Name         string              `json:"name"`
+	Login        string              `json:"login"`
+	Scopes       []string            `json:"scope"`
+	Entitlements map[string][]string `json:"entitlements"`
+}
+
+func (s *ExtendedJWT) Authenticate(ctx context.Context, r *authn.Request) (*authn.Identity, error) {
+	jwtToken := s.retrieveToken(r.HTTPRequest)
+
+	claims, err := s.verifyRFC9068Token(ctx, jwtToken)
+	if err != nil {
+		s.log.Error("Failed to verify JWT", "error", err)
+		return nil, errJWTInvalid.Errorf("Failed to verify JWT: %w", err)
+	}
+
+	// user:id:18
+	userID, err := strconv.ParseInt(strings.TrimPrefix(claims.Subject, fmt.Sprintf("%s:id:", authn.NamespaceUser)), 10, 64)
+	if err != nil {
+		s.log.Error("Failed to parse sub", "error", err)
+		return nil, errJWTInvalid.Errorf("Failed to parse sub: %w", err)
+	}
+
+	if r.OrgID != s.getDefaultOrgID() {
+		s.log.Error("Failed to verify the Organization: OrgID is not the default")
+		return nil, errJWTInvalid.Errorf("Failed to verify the Organization. Only the default org is supported")
+	}
+
+	signedInUser, err := s.userService.GetSignedInUserWithCacheCtx(ctx, &user.GetSignedInUserQuery{OrgID: s.getDefaultOrgID(), UserID: userID})
+	if err != nil {
+		s.log.Error("Failed to get user", "error", err)
+		return nil, errJWTInvalid.Errorf("Failed to get user: %w", err)
+	}
+
+	if signedInUser.Permissions == nil {
+		signedInUser.Permissions = make(map[int64]map[string][]string)
+	}
+
+	if len(claims.Entitlements) == 0 {
+		s.log.Error("Entitlements claim is missing")
+		return nil, errJWTInvalid.Errorf("Entitlements claim is missing")
+	}
```

Comment:
Nit. Should this be moved to `verifyRFC9068Token`?

Reply:
Nope: the aim of the `verifyRFC9068Token` is to verify the token is formally correct. Having a non-emtpy `entitlements` claim is required by our "business-logic" and it is not demanded by the RFC standard thus I put it into the `Authenticate` function.